actionpack/lib/action_dispatch/request/utils.rb

```diff
@@ -33,7 +33,7 @@ def self.check_param_encoding(params)
         when Array
           params.each { |element| check_param_encoding(element) }
         when Hash
-          params.each_value { |value| check_param_encoding(value) }
+          params.each { |value| check_param_encoding(value) }
```

Comment:
Faster implementation, either because it avoids array allocations or doesn't recurse as much (or both). If performance is even more of a concern, check out `proposed_2` in the benchmarking script, though it doesn't look as good anymore.

```suggestion
          params.each do |key, value|
            check_param_encoding(key)
            check_param_encoding(value)
          end
```

```rb
require "benchmark/ips"

def check_param_encoding(params)
  case params
  when Array
    params.each { |element| check_param_encoding(element) }
  when Hash
    params.each_value { |value| check_param_encoding(value) }
  when String
    unless params.valid_encoding?
      raise Rack::Utils::InvalidParameterError, "Invalid encoding for parameter: #{params.scrub}"
    end
  end
end

def check_param_encoding_pr(params)
  case params
  when Array
    params.each { |element| check_param_encoding(element) }
  when Hash
    params.each { |value| check_param_encoding(value) }
  when String
    unless params.valid_encoding?
      raise Rack::Utils::InvalidParameterError, "Invalid encoding for parameter: #{params.scrub}"
    end
  end
end

def check_param_encoding_proposed(params)
  case params
  when Array
    params.each { |element| check_param_encoding(element) }
  when Hash
    params.each do |key, value|
      check_param_encoding(key)
      check_param_encoding(value)
    end
  when String
    unless params.valid_encoding?
      raise Rack::Utils::InvalidParameterError, "Invalid encoding for parameter: #{params.scrub}"
    end
  end
end

def check_param_encoding_proposed_2(params)
  case params
  when Array
    params.each { |element| check_param_encoding(element) }
  when Hash
    params.each do |key, value|
      unless key.valid_encoding?
        raise Rack::Utils::InvalidParameterError, "Invalid encoding for parameter: #{key.scrub}"
      end
      check_param_encoding(value)
    end
  when String
    unless params.valid_encoding?
      raise Rack::Utils::InvalidParameterError, "Invalid encoding for parameter: #{params.scrub}"
    end
  end
end

params = { "example" => "params", "for" => ["1", "2", "3"], "benchmarking" => { "purposes" => "foo", "bar" => "baz" } }

Benchmark.ips do |x|
  x.report("current") { check_param_encoding(params) }
  x.report("pr") { check_param_encoding_pr(params) }
  x.report("proposed") { check_param_encoding_proposed(params) }
  x.report("proposed_2") { check_param_encoding_proposed_2(params) }
  
  x.compare!
end
```

```
ruby 3.3.1 (2024-04-23 revision c56cd86388) [x86_64-linux]
Warming up --------------------------------------
             current   110.359k i/100ms
                  pr    64.145k i/100ms
            proposed    83.416k i/100ms
          proposed_2   106.491k i/100ms
Calculating -------------------------------------
             current      1.112M (± 0.6%) i/s -      5.628M in   5.062857s
                  pr    648.208k (± 0.3%) i/s -      3.271M in   5.046882s
            proposed    836.870k (± 2.0%) i/s -      4.254M in   5.085633s
          proposed_2      1.053M (± 1.0%) i/s -      5.325M in   5.059225s

Comparison:
             current:  1111722.8 i/s
          proposed_2:  1052548.3 i/s - 1.06x  slower
            proposed:   836870.2 i/s - 1.33x  slower
                  pr:   648208.0 i/s - 1.72x  slower
```

Reply:
Cool thanks, did not crossed my mind to do this test. I like the `proposed` more because it does not duplicate logic